public/app/plugins/datasource/cloudwatch/datasource.js

```diff
@@ -416,7 +416,16 @@ function (angular, _, moment, dateMath, kbn, templatingVariable, CloudWatchAnnot
       return _.chain(targets)
       .map(function(target) {
         var dimensionKey = _.findKey(target.dimensions, function(v) {
-          return templateSrv.variableExists(v) && !_.has(scopedVars, templateSrv.getVariableName(v));
+          if (!templateSrv.variableExists(v)) {
+            return false;
+          }
+
+          templateSrv.getVariableNames(v).forEach(function(variableName) {
+            if (_.has(scopedVars, variableName)) {
+              return false;
+            }
+          });
+          return true;
```

Comment:
> Was a mistake to let cloudwatch use templating like it was never designed for.

I apologize for adding this feature. This feature is very hacky, bad design, but very attractive for AWS user...

> loop over all dimensions, find the first one that contains a variable, but not if the variable exist in local scope.

Your regonition is correct.

> why only find the first dimension that matches criteria?

It is complicated to expand variable in all dimension.
I didn't to support it. Only support one dimension expansion.

To expand variable for CloudWatch datasource, we need to find the target CloudWatch dimension, which include template variables, but not include row/panel repeat variable.

```
if (!templateSrv.variableExists(v)) {
  return false;
}
```
Skip the dimension which doesn't include variable.

```
templateSrv.getVariableNames(v).forEach(function(variableName) {
  if (_.has(scopedVars, variableName)) {
    return false;
  }
});
```
Check whether all variable in dimValue which isn't row/panel repeat variable.

> this code looks like it does nothing? loops over variables and then return true or false, but since the return is inside a function this is actually doing nothing.

This is my mistake. Sorry again, fix it...